torch/utils/data/_typing.py

```diff
@@ -141,3 +154,143 @@ def _issubtype_with_constraints(variant, constraints):
+    """
+    type: _DataPipeType
+
+    def __new__(cls, name, bases, namespace, **kargs):
+        # For Python > 3.6
+        cls.__origin__ = None
+        if 'type' in namespace:
+            return super().__new__(cls, name, bases, namespace)
+
+        # For plain derived class without annotation
+        for base in bases:
+            if isinstance(base, _DataPipeMeta):
+                return super().__new__(cls, name, bases, namespace)
+
+        namespace.update({'type': _DEFAULT_TYPE, '__init_subclass__': _dp_init_subclass})
+        return super().__new__(cls, name, bases, namespace)
+
+    @_tp_cache
+    def __getitem__(self, param):
+        if param is None:
+            raise TypeError('{}[t]: t can not be None'.format(self.__name__))
+        if isinstance(param, Sequence):
+            param = Tuple[param]
+        _type_check(param, msg="{}[t]: t must be a type".format(self.__name__))
+        t = _DataPipeType(param)
+
+        if not t.issubtype(self.type):
+            raise TypeError('Can not subclass a DataPipe[{}] from DataPipe[{}]'
+                            .format(t, self.type))
+
+        # Types are equal, fast path for inheritance
+        if self.type.issubtype(t):
+            if _mro_subclass_init(self):
+                return self
+
+        name = self.__name__ + '[' + str(t) + ']'
+        bases = (self,) + self.__bases__
+
+        return self.__class__(name, bases,
+                              {'__init_subclass__': _dp_init_subclass,
+                               'type': t})
+
+    def __eq__(self, other):
+        if not isinstance(other, _DataPipeMeta):
+            return NotImplemented
+        if self.__origin__ is None or other.__origin__ is None:
+            return self is other
+        return (self.__origin__ == other.__origin__
+                and self.type == other.type)
+
+    def __hash__(self):
+        return hash((self.__name__, self.type))
+
+
+def _mro_subclass_init(obj):
+    r"""
+    Run through MRO to check if any super class has already built in
+    the corresponding `__init_subclass__`. If so, no need to add
+    `__init_subclass__`.
+    """
+
+    mro = obj.__mro__
+    for b in mro:
+        if isinstance(b, _DataPipeMeta):
+            if b.__init_subclass__ == _dp_init_subclass:
+                return True
+            if hasattr(b.__init_subclass__, '__func__') and \
+                    b.__init_subclass__.__func__ == _dp_init_subclass:  # type: ignore
+                return True
+    return False
+
+
+def _dp_init_subclass(sub_cls, *args, **kwargs):
+    # TODO:
+    # - add global switch for type checking at compile-time
+    if '__iter__' in sub_cls.__dict__:
+        iter_fn = sub_cls.__dict__['__iter__']
+        try:
+            hints = get_type_hints(iter_fn)
+        # Ignore TypeError causes by invalid string typing annotation in order to keep internal BC
+        except TypeError:
+            return
```

Comment:
There are bugs of typing in internal codebase, have to ignore the exception for non-BC breaking

Reply:
Please file issue for it and add warning.